added json support to env var configuration injection

eeyun opened the pull request:
Ok, this is a tough one and at first glance it feelsbadman. But this is the outcome of https://github.com/habitat-sh/habitat/issues/1938 . For history and explanation you can read read the lengthy backscroll on the issue.

That being said for the sake of anyone following along and for the sake of historical context I'm going to put the TL;DR here:

12 Factor application patterns include a specific API for application configuration which is the ENVVVAR. As habitat has adopted a design that makes the 12 Factor approach to apps easier we've included a granular way to  configure a habitat package via a habitat prefixed environment variable `HAB_PKGNAME`. 

As users have started to discover more workflow and patterns around building and testing their apps as habitat packages some pain points have arrived around this API. What we've discovered is that there's some real pain around utilizing it to do more than set a single value. The reason being the syntax is obtuse and not easily read due to only accepting TOML as our configuration values. TOML is excellently readable in a file format, but as it is line-break delimited injecting configuration values here feels obtuse and somewhat arcane.

This problem gets exacerbated further by other developer and ops tooling around the package. The prime example being docker-compose which is a TML format and doesn't support multi-line envvars. You run into similar issues with things like kubernetes pod definitions. There are workarounds in that you can mount a file that contains your configuration into each container that you run but this adds to a negative experience for first-time users who are expecting a more natural ENVVAR api.

Through lots of discussion on the above RFC I think that we've come to the conclusion that providing JSON as an optional input type for this specific API is the least gross method for correcting the problem currently.

Signed-off-by: eeyun <ihenry@chef.io>

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

mwrock commented on components/sup/src/manager/service/config.rs:
```diff
@@ -371,6 +380,14 @@ fn is_toml_value_a_table(key: &str, table: &toml::value::Table) -> bool {
     }
 }
 
+// fn is_environment_toml(config: &str) -> bool {
```
I'm assuming this is something that got left by accident.

eeyun replied:
oh whoops! let me clean that up real quick.

eeyun commented:
I'm going to correct the commented nonsense and add docs to the PR before we merge. Going to add a DO NOT MERGE flag in the meantime

eeyun commented:
@mwrock updated  to include docs as well as removed the errant commented function

eeyun commented:
I must have the wrong version of rustfmt or something. Any idea what version we're supposed to be using?

eeyun commented:
Corrected Rustfmt issue

reset commented on components/sup/src/manager/service/config.rs:
```diff
@@ -174,11 +175,20 @@ impl Cfg {
+                match toml::de::from_str(&config) {
+                    Ok(toml) => self.environment = Some(toml::Value::Table(toml)),
+                    Err(e) => {
+                        outputln!("{}", sup_error!(Error::TomlParser(e)));
```
I wouldn't output that there was a parser error here since we're going to try and fallback to `JSON`. We can think of this instead like a "hint" system where we aren't sure what your intent was - we don't want to warn or error because we don't actually know what the intent was.

I would remove the Json/Toml parser errors here entirely and instead use early returns to short circuit the function. If we don't early return out, have the last line of this return an error like `Error::BadEnvConfig` which states that we were unable to read valid TOML or JSON and print out the contents that we did receive. You can store the contents in the enum variant like this `Error::BadEnvConfig(String)`

reset commented on components/sup/src/error.rs:
```diff
@@ -505,6 +508,12 @@ impl From<toml::de::Error> for SupError {
     }
 }
 
+impl From<serde_json::Error> for SupError {
```
Per my comment below, I'd remove this trait implementation and the JsonParser error completely

eeyun commented:
@reset i've gone ahead and cleaned up the commit history and added changes in the vein of what you asked before. Let me know if theres more that I should do or if I should change this implementation

reset commented on components/sup/src/error.rs:
```diff
@@ -159,6 +160,7 @@ pub enum Error {
     TomlEncode(toml::ser::Error),
     TomlMergeError(String),
     TomlParser(toml::de::Error),
+    JsonParser(serde_json::Error),
```
We don't need this error anymore

reset commented on components/sup/src/manager/service/config.rs:
```diff
@@ -174,11 +175,20 @@ impl Cfg {
-                })?;
-                self.environment = Some(toml::Value::Table(toml));
-
+                match toml::de::from_str(&config) {
```
The "short circuit"/"early return" pattern I was thinking of would make this function look a bit like this:

```rust
fn load_environment(&mut self, package: &Pkg) -> Result<()> {
    let var_name = format!("{}_{}", ENV_VAR_PREFIX, package.name)
        .to_ascii_uppercase()
        .replace("-", "_");
    match env::var(&var_name) {
        Ok(config) => {
            match toml::de::from_str(&config) {
                Ok(toml) => {
                    self.environment = Some(toml::Value::Table(toml));
                    return Ok(());
                },
                Err(err) => debug!("Attempted to parse env config as toml and failed, {}", err),
            }
            match serde_json::from_str(&config) {
                Ok(json) => {
                    self.environment = Some(toml::Value::Table(json));
                    return Ok(());
                },
                Err(err) => debug!("Attempted to parse env config as json and failed, {}", err),
            }
            self.environment = None;
            Err(Error::BadEnvConfig(package.name.to_ascii_uppercase()))
        }
        Err(e) => {
            debug!(
                "Looking up environment variable {} failed: {:?}",
                var_name,
                e
            );
            self.environment = None;
            Ok(())
        }
    }
}
```

I recommend this pattern for this function becuase it let's you "fall through" and then eventually fail

eeyun commented:
Per @reset in slack. Changes implemented self merge iminent

![](https://media.giphy.com/media/DKnMqdm9i980E/giphy-downsized.gif)

eeyun commented:
@thesentinels approve

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/263135756).

thesentinels commented:
:broken_heart: Travis CI reports [this PR failed to pass the test suite](https://travis-ci.org/habitat-sh/habitat/builds/263135756).

The next step is to examine the [job](https://travis-ci.org/habitat-sh/habitat/builds/263135756) and figure out why. If it is transient, you can try re-triggering the Travis CI Job - if it passes, this PR will be automatically merged. If it is not transient, you should fix the issue and update this pull request, and issue `approve` again. If you believe it will never pass, and you are feeling :godmode:, you can issue a `force` to merge this PR anyway.

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/263135756).

thesentinels commented:
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/263135756).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

![gif-keyboard-3280869874741411265](https://cloud.githubusercontent.com/assets/4304/17755674/1e577b82-6490-11e6-96b7-1663d283c824.gif)

scotthain commented:
YAAAZZZ!!! 😍
